File: 2024/day14/python/solution.py
from pathlib import Path
import re

input_text = (Path(__file__).parent.parent / "input.txt").read_text().strip()

WIDTH = 101
HEIGHT = 103


def parse_robots(text: str) -> list[tuple[int, int, int, int]]:
    """Parse robot positions and velocities."""
    robots = []
    for line in text.split("\n"):
        match = re.match(r"p=(-?\d+),(-?\d+) v=(-?\d+),(-?\d+)", line)
        if match:
            px, py, vx, vy = map(int, match.groups())
            robots.append((px, py, vx, vy))
    return robots


def simulate(robots: list[tuple], seconds: int) -> list[tuple[int, int]]:
    """Simulate robot movement for given seconds."""
    positions = []
    for px, py, vx, vy in robots:
        # Position after 'seconds' time, with wrapping
        new_x = (px + vx * seconds) % WIDTH
        new_y = (py + vy * seconds) % HEIGHT
        positions.append((new_x, new_y))
    return positions


def count_quadrants(positions: list[tuple[int, int]]) -> tuple[int, int, int, int]:
    """Count robots in each quadrant, excluding middle row/column."""
    mid_x = WIDTH // 2   # 50
    mid_y = HEIGHT // 2  # 51

    q1 = q2 = q3 = q4 = 0

    for x, y in positions:
        if x == mid_x or y == mid_y:
            continue  # Skip robots on middle lines

        if x < mid_x and y < mid_y:
            q1 += 1  # Top-left
        elif x > mid_x and y < mid_y:
            q2 += 1  # Top-right
        elif x < mid_x and y > mid_y:
            q3 += 1  # Bottom-left
        else:
            q4 += 1  # Bottom-right

    return q1, q2, q3, q4


def part1() -> int:
    """Part 1: Safety factor after 100 seconds."""
    robots = parse_robots(input_text)
    positions = simulate(robots, 100)
    q1, q2, q3, q4 = count_quadrants(positions)
    return q1 * q2 * q3 * q4


def part2() -> int:
    """Part 2: Find when robots form a Christmas tree pattern."""
    robots = parse_robots(input_text)

    # The Christmas tree appears when robots cluster together
    # Look for a frame with a long horizontal line of robots (tree base/border)
    for seconds in range(1, WIDTH * HEIGHT + 1):
        positions = simulate(robots, seconds)
        pos_set = set(positions)

        # Look for a horizontal line of at least 20 consecutive robots
        for y in range(HEIGHT):
            max_consecutive = 0
            consecutive = 0
            for x in range(WIDTH):
                if (x, y) in pos_set:
                    consecutive += 1
                    max_consecutive = max(max_consecutive, consecutive)
                else:
                    consecutive = 0

            if max_consecutive >= 20:
                return seconds

    return -1


if __name__ == "__main__":
    print(f"Part 1: {part1()}")
    print(f"Part 2: {part2()}")

File: 2024/day14/ruby/solution.rb
#!/usr/bin/env ruby

require 'pathname'
require 'set'

input_text = Pathname.new(__FILE__).dirname.parent.join('input.txt').read.strip

WIDTH = 101
HEIGHT = 103

# Parse robot positions and velocities
def parse_robots(text)
  robots = []
  text.each_line do |line|
    if line =~ /p=(-?\d+),(-?\d+) v=(-?\d+),(-?\d+)/
      px, py, vx, vy = $1.to_i, $2.to_i, $3.to_i, $4.to_i
      robots << [px, py, vx, vy]
    end
  end
  robots
end

# Simulate robot movement for given seconds
def simulate(robots, seconds)
  positions = []
  robots.each do |px, py, vx, vy|
    # Position after 'seconds' time, with wrapping
    # Ruby's % handles negative numbers correctly
    new_x = (px + vx * seconds) % WIDTH
    new_y = (py + vy * seconds) % HEIGHT
    positions << [new_x, new_y]
  end
  positions
end

# Count robots in each quadrant, excluding middle row/column
def count_quadrants(positions)
  mid_x = WIDTH / 2   # 50
  mid_y = HEIGHT / 2  # 51

  q1 = q2 = q3 = q4 = 0

  positions.each do |x, y|
    next if x == mid_x || y == mid_y  # Skip robots on middle lines

    if x < mid_x && y < mid_y
      q1 += 1  # Top-left
    elsif x > mid_x && y < mid_y
      q2 += 1  # Top-right
    elsif x < mid_x && y > mid_y
      q3 += 1  # Bottom-left
    else
      q4 += 1  # Bottom-right
    end
  end

  [q1, q2, q3, q4]
end

# Part 1: Safety factor after 100 seconds
def part1(robots)
  positions = simulate(robots, 100)
  q1, q2, q3, q4 = count_quadrants(positions)
  q1 * q2 * q3 * q4
end

# Part 2: Find when robots form a Christmas tree pattern
def part2(robots)
  # The Christmas tree appears when robots cluster together
  # Look for a frame with a long horizontal line of robots (tree base/border)
  (1..(WIDTH * HEIGHT)).each do |seconds|
    positions = simulate(robots, seconds)
    pos_set = positions.to_set

    # Look for a horizontal line of at least 20 consecutive robots
    HEIGHT.times do |y|
      max_consecutive = 0
      consecutive = 0

      WIDTH.times do |x|
        if pos_set.include?([x, y])
          consecutive += 1
          max_consecutive = [max_consecutive, consecutive].max
        else
          consecutive = 0
        end
      end

      return seconds if max_consecutive >= 20
    end
  end

  -1
end

if __FILE__ == $0
  robots = parse_robots(input_text)
  puts "Part 1: #{part1(robots)}"
  puts "Part 2: #{part2(robots)}"
end
